### PR TITLE
Address a number of IOperation issues for VB

### LINKED ISSUE #22066: Incorrect ArgumentKind for argument to query invocation
Test `ImplicitSelectClause_IOperation` in QueryExpressions.vb
See https://github.com/dotnet/roslyn/pull/21356#issuecomment-325050215

### LINKED ISSUE #22486: Ensure correct syntax node is associated with matched BoundOmittedArgument
See unit test `DirectlyBindOmittedArgument_InvocationExpression` and the baseline operation tree for the test. Issue is explained here: https://github.com/dotnet/roslyn/pull/22444#discussion_r142282133 and the underlying reason is explained here: https://github.com/dotnet/roslyn/pull/22444#discussion_r142439183

### PR DESCRIPTION
**Pull request by AlekseyTs**
1. Ensure correct syntax nodes and implicit state are used for IArgument nodes matching OmittedArgumentSyntax. Fixes #22486.
2. Ensure correct argument kind is associated with synthesized arguments within query expressions. Fixes #22066.
3. Ensure the invariant that “each syntax node has at most one corresponding explicit IOperation node” is met for all test scenarios we have at the moment. In order to achieve this the following changes are made:
-mark some synthesized bound nodes as compiler generated.
-for assignment statements, ensure that ISimpleAssignmentExpression and ICompoundAssignmentExpression nodes are marked as Implicit because they are sharing syntax node with containing IExpressionStatement and VB doesn’t have a concept of an assignment expression.
-for throw statements, ensure that IThrowExpression node is marked as Implicit because it is sharing syntax node with containing IExpressionStatement and VB doesn’t have a concept of a throw expression.
-for Add/RemoveHandler statements, ensure that IEventAssignmentExpression node is marked as Implicit because it is sharing syntax node with containing IExpressionStatement and VB doesn’t have a concept of an event assignment expression.
-ensure that receivers for IPropertyReferenceExpression representing references to query range variables are marked as implicit.
-ensure that IPropertyReferenceExpression and IParameterReferenceExpression nodes representing references to query range variables are associated with correct syntax nodes.
-ensure that return statements and blocks in query lambdas are marked as implicit.
-for IInterpolatedStringText node, ensure the underlying ILiteralExpression is marked as implicit because it is sharing syntax node with its parent.

VSO bug : https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/512836

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @dotnet/analyzer-ioperation, @cston Please review 15.5 fix.

**Comment by AlekseyTs** [comment]
CC @jinujoseph For ask mode approval

**Comment by jcouv** [comment]
Looking

**Comment by jcouv** [comment]
>             Return New BoundExpressionStatement(statement, boundExpression)

Should `boundExpression` be marked as compiler-generated here?

---
Refers to: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb:2582 in 9fed8fa. [](commit_id = 9fed8fa02cd865b02ce81e99bfb56ad7eabe6cbc, deletion_comment = False)

**Comment by jcouv** [comment]
>             Return New BoundExpressionStatement(printStmt, boundExpression)

Same question here and a couple of lines below.

---
Refers to: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb:2656 in 9fed8fa. [](commit_id = 9fed8fa02cd865b02ce81e99bfb56ad7eabe6cbc, deletion_comment = False)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.vb [hunk 1]
```diff
@@ -479,18 +479,18 @@ IBlockStatement (26 statements, 3 locals) (OperationKind.BlockStatement) (Syntax
             Operand: 
               IBinaryOperatorExpression (BinaryOperatorKind.Concatenate, Checked) (OperationKind.BinaryOperatorExpression, Type: System.String) (Syntax: 'x & y')
                 Left: 
-                  IConversionExpression (Explicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.String) (Syntax: 'x')
+                  IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.String, IsImplicit) (Syntax: 'x')
```
Comment (agocke) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.vb]: Is this a consequence of the `ISimpleAssignmentExpression` being `Implicit`? i.e., `Implicit`ness is transitive in some sense?

That seems fine... but I'm confused as to why this was classified as an explicit conversion to begin with. Where's the explicit conversion in `x & y`?

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 2]
```diff
@@ -252,6 +259,18 @@ public override void Visit(IOperation operation)
...
 
+            if (_explictNodeMap != null && !operation.IsImplicit)
+            {
+                try
+                {
+                    _explictNodeMap.Add(operation.Syntax, operation);
+                }
+                catch (ArgumentException)
+                {
+                    Assert.False(true, $"Duplicate explicit node for syntax ({operation.Syntax.RawKind}): {operation.Syntax.ToString()}");
```
Comment (agocke) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: I love this -- thank you for adding this check.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 3]
```diff
@@ -44,7 +50,8 @@ public static void Verify(Compilation compilation, IOperation operation, string
 
         public static string GetOperationTree(Compilation compilation, IOperation operation, int initialIndent = 0)
         {
-            var walker = new OperationTreeVerifier(compilation, operation, initialIndent);
+            var walker = new OperationTreeVerifier(compilation, operation, initialIndent,
```
Comment (agocke) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: Is this run in both Debug and Release modes?

**Review by agocke — APPROVED** [review]
LGTM, just a few questions.

File: src/Compilers/VisualBasic/Portable/BoundTree/BoundInterpolatedStringExpression.vb [hunk 4]
```diff
@@ -36,6 +36,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 #If DEBUG Then
         Private Sub Validate()
             Debug.Assert(Type.SpecialType = SpecialType.System_String)
+            Debug.Assert(Not Contents.Where(Function(content) content.Kind <> BoundKind.Interpolation AndAlso content.Kind <> BoundKind.Literal).Any())
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/BoundTree/BoundInterpolatedStringExpression.vb]: Consider simplifying: `Contents.All(Function(content) content.Kind = Interpolation OrElse content.Kind = Literal)`

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.vb (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.vb]: > Is this a consequence of the ISimpleAssignmentExpression being Implicit? i.e., Implicitness is transitive in some sense?

No, this isn't related to ISimpleAssignmentExpression node at all. Implicit conversions within concatenate operator aren't supposed to produce any diagnostics originated from the fact that they are implicit. And that was leaking to the IConversionExpression node, see the logic around ```explicitSemanticForConcatArgument``` in Binder_Operators.vb.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 5]
```diff
@@ -44,7 +50,8 @@ public static void Verify(Compilation compilation, IOperation operation, string
 
         public static string GetOperationTree(Compilation compilation, IOperation operation, int initialIndent = 0)
         {
-            var walker = new OperationTreeVerifier(compilation, operation, initialIndent);
+            var walker = new OperationTreeVerifier(compilation, operation, initialIndent, 
+                trackExplicitNodes: operation.Language == LanguageNames.VisualBasic);
```
Comment (jcouv) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: I assume we'll apply the same invariant to IOperation in C# at some point. Is that correct?

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: > Is this run in both Debug and Release modes?

I believe so.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Operators.vb [hunk 6]
```diff
@@ -446,9 +446,17 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+            Dim beforeConversion As BoundExpression = left
             left = ApplyConversion(left.Syntax, operandType, left, explicitSemanticForConcatArgument, diagnostics,
                                    explicitSemanticForConcatArgument:=explicitSemanticForConcatArgument)
 
+            If explicitSemanticForConcatArgument AndAlso left IsNot beforeConversion AndAlso left.Kind = BoundKind.Conversion Then
+                Dim conversion = DirectCast(left, BoundConversion)
+                left = conversion.Update(conversion.Operand, conversion.ConversionKind, conversion.Checked, explicitCastInCode:=False,
+                                         constantValueOpt:=conversion.ConstantValueOpt, extendedInfoOpt:=conversion.ExtendedInfoOpt,
+                                         type:=conversion.Type)
+            End If
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Operators.vb]: I'd be tempted to extract both blocks as helper method.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: > I assume we'll apply the same invariant to IOperation in C# at some point. Is that correct?

Yes, the invariant is applicable to both languages.

**Comment by AlekseyTs** [comment]
>             Return New BoundExpressionStatement(statement, boundExpression)

> Should boundExpression be marked as compiler-generated here?

No, the statement and expression are distinct syntax nodes.

---
In reply to: [338750530](https://github.com/dotnet/roslyn/pull/22795#issuecomment-338750530) [](ancestors = 338750530)

---
Refers to: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb:2582 in 9fed8fa. [](commit_id = 9fed8fa02cd865b02ce81e99bfb56ad7eabe6cbc, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>             Return New BoundExpressionStatement(printStmt, boundExpression)

Same answer. 

---
In reply to: [338750699](https://github.com/dotnet/roslyn/pull/22795#issuecomment-338750699) [](ancestors = 338750699)

---
Refers to: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb:2656 in 9fed8fa. [](commit_id = 9fed8fa02cd865b02ce81e99bfb56ad7eabe6cbc, deletion_comment = False)

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 7]
```diff
@@ -1479,7 +1479,7 @@ Namespace Microsoft.CodeAnalysis.Semantics
             If boundNode.Kind = BoundKind.Interpolation Then
                 Return DirectCast(Create(boundNode), IInterpolatedStringContent)
             Else
-                Return CreateBoundInterpolatedStringTextOperation(boundNode)
+                Return CreateBoundInterpolatedStringTextOperation(DirectCast(boundNode, BoundLiteral))
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: I didn't understand the purpose of the cast. There seems to only be one `CreateBoundInterpolatedStringTextOperation` overload.

**Review by jcouv — APPROVED** [review]
LGTM

**Comment by natidea** [comment]
Approved to merge via [512836](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/512836)

**Comment by AlekseyTs** [comment]
Ping. @dotnet/analyzer-ioperation Please review 15.5 fix. That is the only thing blocking the merge .

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 8]
```diff
@@ -1206,7 +1206,7 @@ IVariableDeclarationStatement (1 declarations) (OperationKind.VariableDeclaratio
     Variables: Local_1: x As System.Int32()
     Initializer: 
       IVariableInitializer (OperationKind.VariableInitializer, IsImplicit) (Syntax: 'x(2)')
-        IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: System.Int32()) (Syntax: 'x(2)')
+        IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: System.Int32(), IsImplicit) (Syntax: 'x(2)')
```
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: It feels a little weird to me that the ArrayCreation is the implicit node. I'd expect the VariableInitializer to be the implicit one.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 5 above)
Comment (333fred) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: Then why aren't we currently doing both languages? Is there a separate issue tracking that?

---
In reply to: [146360593](https://github.com/dotnet/roslyn/pull/22795#discussion_r146360593) [](ancestors = 146360593)

Comment (333fred) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: Is there a separate issue tracking enabling this for C# then?

**Review by 333fred — APPROVED** [review]
Looks good to me. My comments can be addressed by followups.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: > Is there a separate issue tracking enabling this for C# then?

There is an existing issue about this - https://github.com/dotnet/roslyn/issues/22229. I am simply volunteering to do this work.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb (hunk 8 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: > It feels a little weird to me that the ArrayCreation is the implicit node. I'd expect the VariableInitializer to be the implicit one.

I believe all nodes are supposed to be implicit here except the size of the array. There is no code that explicitly creates an array.

File: src/Compilers/VisualBasic/Portable/BoundTree/BoundInterpolatedStringExpression.vb (hunk 4 above)
Comment (jcouv) [reply src/Compilers/VisualBasic/Portable/BoundTree/BoundInterpolatedStringExpression.vb]: Ping just to make sure you didn't miss this comment. Thanks

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/BoundTree/BoundInterpolatedStringExpression.vb]: > Ping just to make sure you didn't miss this comment.

Opened https://github.com/dotnet/roslyn/issues/22869

**Pull request merged**